### ISSUE TITLE
Rule proposal: `no-unused-return-value`

### ISSUE THREAD

**Issue by silverwind** [sindresorhus/eslint-plugin-unicorn#741]
Details here: https://github.com/eslint/eslint/issues/12980

The idea is that many builtin functions return values and if those return values are not used, the code is essentialy a noop and can be removed. Notably if getters are involved, the rule may have false positives but I generally see getters as a obscure feature that should not be used.

The discussion got stale because it seems hard for a plugin to correctly judge the type of a variable which is essential for a rule like this so it may require Typescript to really be useful.

### Fail
```` js
const arr = [{name: 'eslint'}];
arr.map(item => item.name); // error: Unused Array.prototype.map return value
console.log(arr);
````
### Pass
```` js
const arr = [{name: 'eslint'}];
console.log(arr);
````

**Comment by papb** [comment]
Hmm, I don't like this idea, at least not how I currently understand it. There are so many impure functions whose return value can be safely discarded in several scenarios... Example:

```js
const foo = new Set(['foo']);
foo.delete('foo');
```

This call has a return value, in this case it returns true, but it is very common to simply ignore this.

**Comment by silverwind** [comment]
Yes of course it'd only enable it for carefully selected functions. For Array and Object, I could see

````
Array.isArray
Array.of
Array.prototype.concat
Array.prototype.filter
Array.prototype.includes
Array.prototype.indexOf
Array.prototype.join
Array.prototype.lastIndexOf
Array.prototype.slice
Array.prototype.toSource
Array.prototype.toString
Array.prototype.toLocaleString
Array.prototype.entries
Array.prototype.every
Array.prototype.find
Array.prototype.findIndex
Array.prototype.keys
Array.prototype.map
Array.prototype.some
Array.prototype.values

Object.create
Object.entries
Object.fromEntries
Object.getOwnPropertyDescriptor
Object.getOwnPropertyDescriptors
Object.getOwnPropertyNames
Object.getOwnPropertySymbols
Object.getPrototypeOf
Object.is
Object.isExtensible
Object.isFrozen
Object.isSealed
Object.keys
Object.values
Object.prototype.hasOwnProperty
Object.prototype.isPrototypeOf
Object.prototype.propertyIsEnumerable
Object.prototype.toLocaleString
Object.prototype.toString
````

**Comment by papb** [comment]
Ah!! Of course! Awesome.

**Comment by fisker** [comment]
```js
Function#bind()

Reflect#getOwnPropertyDescriptor()
...

Canvas#toDataUrl()

URL.createObjectURL()

document.createElement()
document.querySelector()
...

```

**Comment by silverwind** [comment]
Another `map` example added to OP which shows a different resolution by using `forEach`.

**Comment by sindresorhus** [comment]
I like the idea, but I agree, we need to selectively pick the methods, since it's hard to statically determine many cases (without type information, that is).

**Comment by sindresorhus** [comment]
This is accepted.

**Comment by brettz9** [comment]
I would like to request that Array's `every` and `some` not be included (unless by option) since they can be used for short-circuiting.

**Comment by fisker** [comment]
I think these cases are both `ExpressionStatement` in AST, how about a rule named `no-restricted-statements`,

forbid `arr.map(item => item.name)` as a `ExpressionStatement`, and a configurable options accepts [esquery](https://github.com/estools/esquery) selector

```js
{
	'unicorn/no-restricted-statements': [
		'error',
		{
			extendDefaultDenyList: false,
			denyList: [
				// To restrict `foo.bar()`
				'CallExpression[callee.property.name="bar"]'
			]
		}
		
	]
}

**Comment by EvgenyOrekhov** [comment]
I believe this is covered by [`sonarjs/no-ignored-return`](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-ignored-return.md).
